### ISSUE TITLE
[TySan] False positive with global structs

### ISSUE THREAD

**Issue by TheLastRar** [llvm/llvm-project#122954]
This was previously reported in the discourse, since TySan has now been merged I'm reposting it on Github

This is the reduced version from [gbMattN](https://discourse.llvm.org/t/reviving-typesanitizer-a-sanitizer-to-catch-type-based-aliasing-violations/66092/22), the original report used a `std::array<float,2>`

Tysan reports a type violation when arr is in the global scope https://godbolt.org/z/TxeaaY5cc

```C
#include <stdio.h>

struct array_type{
    int inner[1];
};

struct array_type arr;

int main() {
  arr.inner[0] = 5;
  return 0;
}
```

TySan currently reports
```
==1==ERROR: TypeSanitizer: type-aliasing-violation on address 0x5d14c0846c5c (pc 0x5d14bfeefb4b bp 0x7ffd3e06b7a0 sp 0x7ffd3e06b730 tid 1)
WRITE of size 4 at 0x5d14c0846c5c with type int accesses an existing object of type array_type
    #0 0x5d14bfeefb4a  (/app/output.s+0x2ab4a)
```

**Comment by gbMattN** [comment]
Ah! Thanks for pointing this out again. This feels like something #120406 should fix but it doesn't.... hmm

TySan sets the shadow memory for globals at the start. `&arr` and `&arr.inner[0]` is the same, so the TD is `array_type` as it was set at the start. The mistake here is that TySan doesn't dig down into the `array_type` TD to see that the first member is an int.

Will certainly give it a look

**Comment by gbMattN** [comment]
Aha. The inner array type is given the incorrect TD. Sounds like a TBAA issue. I think the draft solution to #119615 should be extendable to fix this too

**Comment by tavianator** [comment]
Another reproducer (I assume?)

```c
struct foo {
	void *ptr;
};

struct bar {
	int n;
	struct foo foo;
};

static void bar_init(struct bar *bar, void *ptr) {
	bar->n = 0;
	bar->foo.ptr = 0;
}

struct baz {
	struct bar bar;
};

static void baz_init(struct baz *baz) {
	bar_init(&baz->bar, 0);
}

static struct baz g;

int main(void) {
	baz_init(&g);
	return 0;
}
```

```
==53816==ERROR: TypeSanitizer: type-aliasing-violation on address 0x55decd5fcc08 (pc 0x55deccca5ad7 bp 0x7ffcb6987520 sp 0x7ffcb6986cc8 tid 53816)
WRITE of size 8 at 0x55decd5fcc08 with type any pointer (in bar at offset 8) accesses part of an existing object of type baz that starts at offset -8
    #0 0x55deccca5ad6 in bar_init /home/tavianator/foo.c:12:15
```
